pkg/lib/interaction/nodes/change_password.go

```diff
@@ -0,0 +1,128 @@
+	Stage interaction.AuthenticationStage `json:"stage"`
+}
+
+func (e *EdgeChangePasswordBegin) Instantiate(ctx *interaction.Context, graph *interaction.Graph, rawInput interface{}) (interaction.Node, error) {
+	return &NodeChangePasswordBegin{
+		Stage: e.Stage,
+	}, nil
+}
+
+type NodeChangePasswordBegin struct {
+	Stage interaction.AuthenticationStage `json:"stage"`
+}
+
+func (n *NodeChangePasswordBegin) Prepare(ctx *interaction.Context, graph *interaction.Graph) error {
+	return nil
+}
+
+func (n *NodeChangePasswordBegin) Apply(perform func(eff interaction.Effect) error, graph *interaction.Graph) error {
+	return nil
+}
+
+func (n *NodeChangePasswordBegin) DeriveEdges(graph *interaction.Graph) ([]interaction.Edge, error) {
+	return []interaction.Edge{&EdgeChangePassword{
+		Stage: n.Stage,
+	}}, nil
+}
+
+type InputChangePassword interface {
+	GetNewPassword() string
+}
+
+type EdgeChangePassword struct {
+	Stage interaction.AuthenticationStage
+}
+
+func (e *EdgeChangePassword) Instantiate(ctx *interaction.Context, graph *interaction.Graph, rawInput interface{}) (node interaction.Node, err error) {
+	input, ok := rawInput.(InputChangePassword)
+	if !ok {
+		return nil, interaction.ErrIncompatibleInput
+	}
+
+	newPassword := input.GetNewPassword()
+
+	userID := graph.MustGetUserID()
+	ais, err := ctx.Authenticators.List(
+		userID,
+		authenticator.KeepType(authn.AuthenticatorTypePassword),
+		authenticator.KeepKind(stageToAuthenticatorKind(e.Stage)),
+	)
+	if err != nil {
+		return
+	}
+
+	if len(ais) == 0 {
+		err = ErrNoPassword
+		return
+	}
+
+	if len(ais) != 1 {
+		err = fmt.Errorf("changepassword: detected user %s having more than 1 password", userID)
+		return
+	}
+
+	oldInfo := ais[0]
+	changed, newInfo, err := ctx.Authenticators.WithSecret(oldInfo, newPassword)
+	if err != nil {
+		return
+	}
+
+	newNode := &NodeChangePasswordEnd{
+		Stage:            e.Stage,
+		OldAuthenticator: oldInfo,
+	}
+	if changed {
+		newNode.NewAuthenticator = newInfo
+	}
+
+	node = newNode
+	return node, nil
+}
+
+type NodeChangePasswordEnd struct {
+	Stage            interaction.AuthenticationStage `json:"stage"`
+	OldAuthenticator *authenticator.Info             `json:"old_authenticator"`
+	NewAuthenticator *authenticator.Info             `json:"new_authenticator,omitempty"`
+}
+
+func (n *NodeChangePasswordEnd) Apply(perform func(eff interaction.Effect) error, graph *interaction.Graph) error {
+	return nil
+}
+
+func (n *NodeChangePasswordEnd) Prepare(ctx *interaction.Context, graph *interaction.Graph) error {
+	return nil
+}
+
+func (n *NodeChangePasswordEnd) DeriveEdges(graph *interaction.Graph) ([]interaction.Edge, error) {
+	if n.NewAuthenticator != nil {
+		return []interaction.Edge{
+			&EdgeDoUpdateAuthenticator{
+				Stage:                     n.Stage,
+				AuthenticatorBeforeUpdate: n.OldAuthenticator,
+				AuthenticatorAfterUpdate:  n.NewAuthenticator,
+			},
+		}, nil
+	}
+	// Password was not changed, ends the interaction
+	return nil, nil
```

Comment:
The intent should control the interaction flow, but this makes the node ends the interaction. Perhaps move this logic to `DeriveEdgesForNode`?